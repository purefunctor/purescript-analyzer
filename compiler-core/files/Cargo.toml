[package]
name = "files"
version = "0.1.0"
edition = "2024"

[dependencies]
indexmap = "2.7.1"
la-arena = "0.3.1"
rustc-hash = "2.1.1"
