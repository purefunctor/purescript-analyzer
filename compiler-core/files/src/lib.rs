use core::str;
use std::sync::Arc;

use indexmap::IndexMap;
use la_arena::{Idx, RawIdx};
use rustc_hash::FxBuildHasher;

#[derive(Debug, Default)]
pub struct Files {
    files: IndexMap<Arc<str>, Arc<str>, FxBuildHasher>,
}

pub struct File;

pub type FileId = Idx<File>;

impl Files {
    pub fn insert(&mut self, k: impl Into<Arc<str>>, v: impl Into<Arc<str>>) -> FileId {
        let k = k.into();
        let v = v.into();
        let (index, _) = self.files.insert_full(k, v);
        Idx::from_raw(RawIdx::from_u32(index as u32))
    }

    pub fn id(&self, k: &str) -> Option<FileId> {
        self.files.get_full(k).map(|(index, _, _)| Idx::from_raw(RawIdx::from_u32(index as u32)))
    }

    pub fn path(&self, id: FileId) -> Arc<str> {
        let index = id.into_raw().into_u32() as usize;
        let (path, _) =
            self.files.get_index(index).expect("invariant violated: expected valid FileId");
        Arc::clone(path)
    }

    pub fn content(&self, id: FileId) -> Arc<str> {
        let index = id.into_raw().into_u32() as usize;
        let (_, contents) =
            self.files.get_index(index).expect("invariant violated: expected valid FileId");
        Arc::clone(contents)
    }
}

#[cfg(test)]
mod tests {
    use super::Files;

    #[test]
    fn test_basic() {
        let mut files = Files::default();

        let k = "src/Main.purs";
        let v = "module Main where\n\n";

        let id = files.insert(k, v);

        assert_eq!(files.id(k), Some(id));
        assert_eq!(files.path(id).as_ref(), k);
        assert_eq!(files.content(id).as_ref(), v);
    }
}
