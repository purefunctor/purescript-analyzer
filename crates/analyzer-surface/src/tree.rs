//! A high-level AST for semantic information.
//!
//! The types defined in this file roughly follows the high-level AST for the
//! PureScript programming language. Syntax sugar is computed much later once
//! more information becomes available.

use std::{borrow::Borrow, ops::Index, sync::Arc};

use analyzer_index::{nominal::*, AstId};
use la_arena::{Arena, Idx};
use paste::paste;
use rustc_hash::FxHashMap;
use syntax::{ast, SyntaxNodePtr};

// region: Names

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct ModuleName(Arc<str>);

impl ModuleName {
    pub fn iter(&self) -> impl Iterator<Item = &str> {
        self.0.split('.')
    }

    pub fn from_raw(inner: Arc<str>) -> ModuleName {
        ModuleName(inner)
    }
}

impl AsRef<str> for ModuleName {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl Borrow<str> for ModuleName {
    fn borrow(&self) -> &str {
        &self.0
    }
}

impl Borrow<str> for &ModuleName {
    fn borrow(&self) -> &str {
        &self.0
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Name(Arc<str>);

impl Name {
    pub fn from_raw(inner: Arc<str>) -> Name {
        Name(inner)
    }
}

impl AsRef<str> for Name {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl Borrow<str> for Name {
    fn borrow(&self) -> &str {
        &self.0
    }
}

impl Borrow<str> for &Name {
    fn borrow(&self) -> &str {
        &self.0
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct Qualified<T> {
    pub prefix: Option<ModuleName>,
    pub value: T,
}

// endregion

// region: Module

#[derive(Debug, PartialEq, Eq)]
pub struct Module {
    pub header: ModuleHeader,
    pub imports: ModuleImports,
    pub body: ModuleBody,
}

// endregion

// region: ModuleHeader

#[derive(Debug, PartialEq, Eq)]
pub struct ModuleHeader {
    pub name: ModuleName,
    pub export_list: Option<ExportList>,
}

#[derive(Debug, PartialEq, Eq)]
pub struct ExportList {
    pub items: Vec<ExportItem>,
}

#[derive(Debug, PartialEq, Eq)]
pub enum ExportItem {
    ExportClass(Name),
    ExportType(Name, Option<DataMembers>),
    ExportValue(Name),
}

// endregion

// region: ModuleImports

#[derive(Debug, PartialEq, Eq)]
pub struct ModuleImports {
    pub declarations: Vec<ImportDeclaration>,
}

#[derive(Debug, PartialEq, Eq)]
pub struct ImportDeclaration {
    pub name: ModuleName,
    pub qualified_as: Option<ModuleName>,
    pub import_list: Option<ImportList>,
}

#[derive(Debug, PartialEq, Eq)]
pub struct ImportList {
    pub items: Vec<ImportItem>,
    pub hiding: bool,
}

#[derive(Debug, PartialEq, Eq)]
pub enum ImportItem {
    ImportClass(Name),
    ImportType(Name, Option<DataMembers>),
    ImportValue(Name),
}

#[derive(Debug, PartialEq, Eq)]
pub enum DataMembers {
    DataAll,
    DataEnumerated(Vec<Name>),
}

// endregion

// region: ModuleBody

#[derive(Debug, Default, PartialEq, Eq)]
pub struct ModuleBody {
    pub declarations: Vec<Declaration>,
    pub class_declarations: FxHashMap<ClassGroupId, usize>,
    pub data_declarations: FxHashMap<DataGroupId, usize>,
    pub value_declarations: FxHashMap<ValueGroupId, usize>,
}

impl ModuleBody {
    pub fn class_declaration(&self, class_id: ClassGroupId) -> Option<&ClassDeclaration> {
        self.class_declarations.get(&class_id).map(|index| {
            let declaration = &self.declarations[*index];
            if let Declaration::ClassDeclaration(class_declaration) = declaration {
                class_declaration
            } else {
                unreachable!("impossible: index does not point to a class declaration");
            }
        })
    }

    pub fn data_declaration(&self, data_id: DataGroupId) -> Option<&DataDeclaration> {
        self.data_declarations.get(&data_id).map(|index| {
            let declaration = &self.declarations[*index];
            if let Declaration::DataDeclaration(data_declaration) = declaration {
                data_declaration
            } else {
                unreachable!("impossible: index does not point to a data declaration");
            }
        })
    }

    pub fn value_declaration(&self, value_id: ValueGroupId) -> Option<&ValueDeclaration> {
        self.value_declarations.get(&value_id).map(|index| {
            let declaration = &self.declarations[*index];
            if let Declaration::ValueDeclaration(value_declaration) = declaration {
                value_declaration
            } else {
                unreachable!("impossible: index does not point to a value declaration");
            }
        })
    }

    pub fn iter_class_declarations(&self) -> impl Iterator<Item = &ClassDeclaration> {
        self.declarations.iter().filter_map(|declaration| {
            if let Declaration::ClassDeclaration(data_declaration) = declaration {
                Some(data_declaration)
            } else {
                None
            }
        })
    }

    pub fn iter_data_declarations(&self) -> impl Iterator<Item = &DataDeclaration> {
        self.declarations.iter().filter_map(|declaration| {
            if let Declaration::DataDeclaration(data_declaration) = declaration {
                Some(data_declaration)
            } else {
                None
            }
        })
    }

    pub fn iter_value_declarations(&self) -> impl Iterator<Item = &ValueDeclaration> {
        self.declarations.iter().filter_map(|declaration| {
            if let Declaration::ValueDeclaration(value_declaration) = declaration {
                Some(value_declaration)
            } else {
                None
            }
        })
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum Declaration {
    ClassDeclaration(ClassDeclaration),
    DataDeclaration(DataDeclaration),
    ValueDeclaration(ValueDeclaration),
}

// endregion

// region: DataDeclaration

#[derive(Debug, PartialEq, Eq)]
pub struct DataDeclaration {
    pub id: DataGroupId,
    pub name: Name,
    pub annotation: Option<TypeId>,
    pub variables: Vec<TypeVariable>,
    pub constructors: FxHashMap<AstId<ast::DataConstructor>, DataConstructor>,
}

#[derive(Debug, PartialEq, Eq)]
pub struct DataConstructor {
    pub name: Name,
    pub fields: Vec<TypeId>,
}

// endregion

// region: ClassDeclaration

#[derive(Debug, PartialEq, Eq)]
pub struct ClassDeclaration {
    pub id: ClassGroupId,
    pub name: Name,
    pub signature: Option<TypeId>,
    pub constraints: Vec<TypeId>,
    pub variables: Vec<TypeVariable>,
    pub fundeps: Vec<FunctionalDependency>,
    pub members: FxHashMap<AstId<ast::ClassMember>, ClassMember>,
}

#[derive(Debug, PartialEq, Eq)]
pub enum FunctionalDependency {
    Determined(Vec<Name>),
    Determines(Vec<Name>, Vec<Name>),
}

#[derive(Debug, PartialEq, Eq)]
pub struct ClassMember {
    pub name: Name,
    pub ty: TypeId,
}

// region: ValueDeclaration

#[derive(Debug, PartialEq, Eq)]
pub struct ValueDeclaration {
    pub id: ValueGroupId,
    pub name: Name,
    pub annotation: Option<TypeId>,
    pub equations: Vec<ValueEquation>,
}

#[derive(Debug, PartialEq, Eq)]
pub struct ValueEquation {
    pub binders: Vec<BinderId>,
    pub binding: Binding,
}

#[derive(Debug, PartialEq, Eq)]
pub enum Binding {
    Unconditional { where_expr: WhereExpr },
    Guarded { guarded_exprs: Vec<GuardedExpr> },
}

#[derive(Debug, PartialEq, Eq)]
pub struct WhereExpr {
    pub expr_id: ExprId,
    pub let_bindings: Vec<LetBinding>,
}

#[derive(Debug, PartialEq, Eq)]
pub struct GuardedExpr {
    pub pattern_guards: Vec<PatternGuard>,
    pub where_expr: WhereExpr,
}

#[derive(Debug, PartialEq, Eq)]
pub struct PatternGuard {
    pub binder_id: Option<BinderId>,
    pub expr_id: ExprId,
}

// endregion

// region: Literal

#[derive(Debug, PartialEq, Eq)]
pub enum Literal<I> {
    Array(Vec<I>),
    Record(Vec<RecordItem<I>>),
    Int(usize),
    Number(Arc<str>),
    String(Arc<str>),
    Char(Arc<str>),
    Boolean(bool),
}

#[derive(Debug, PartialEq, Eq)]
pub enum IntOrNumber {
    Int(usize),
    Number(Arc<str>),
}

#[derive(Debug, PartialEq, Eq)]
pub enum RecordItem<I> {
    RecordPun(Name),
    RecordField(Name, I),
}

// endregion

// region: Expr

#[derive(Debug, PartialEq, Eq)]
pub enum Expr {
    Application(ExprId, Vec<ExprId>),
    Constructor(Qualified<Name>),
    IfThenElse(ExprId, ExprId, ExprId),
    InfixChain(ExprId, Vec<(ExprId, ExprId)>),
    Lambda(Vec<BinderId>, ExprId),
    LetIn(Vec<LetBinding>, ExprId),
    Literal(Literal<ExprId>),
    OperatorChain(ExprId, Vec<(Qualified<Name>, ExprId)>),
    Parenthesized(ExprId),
    Variable(Qualified<Name>),
    NotImplemented,
}

pub type ExprId = Idx<Expr>;

#[derive(Debug, PartialEq, Eq)]
pub enum LetBinding {
    Name { id: LetNameId },
    Pattern { binder: BinderId, where_expr: WhereExpr },
}

#[derive(Debug, PartialEq, Eq)]
pub struct LetName {
    pub name: Name,
    pub annotation: Option<TypeId>,
    pub equations: Vec<LetNameEquation>,
}

pub type LetNameId = Idx<LetName>;

#[derive(Debug, PartialEq, Eq)]
pub struct LetNameEquation {
    pub binders: Vec<BinderId>,
    pub binding: Binding,
}

// endregion

// region: Binder

#[derive(Debug, PartialEq, Eq)]
pub enum Binder {
    Constructor { name: Qualified<Name>, fields: Vec<BinderId> },
    Literal(Literal<BinderId>),
    Negative(IntOrNumber),
    Parenthesized(BinderId),
    Variable(Name),
    Wildcard,
    NotImplemented,
}

pub type BinderId = Idx<Binder>;

// endregion

// region: Type

#[derive(Debug, PartialEq, Eq)]
pub enum Type {
    Arrow(Vec<TypeId>, TypeId),
    Application(TypeId, Vec<TypeId>),
    Constrained(TypeId, TypeId),
    Constructor(Qualified<Name>),
    Forall(Vec<TypeVariable>, TypeId),
    OperatorChain(TypeId, Vec<(Qualified<Name>, TypeId)>),
    Parenthesized(TypeId),
    Variable(Name),
    NotImplemented,
}

pub type TypeId = Idx<Type>;

#[derive(Debug, PartialEq, Eq)]
pub enum TypeVariable {
    Kinded(Name, TypeId),
    Name(Name),
}

impl TypeVariable {
    pub fn to_name(&self) -> Name {
        match self {
            TypeVariable::Kinded(name, _) => Name::clone(name),
            TypeVariable::Name(name) => Name::clone(name),
        }
    }
}

// endregion

// region: SurfaceArena

macro_rules! _surface_arena {
    ($($name:ident: $tree:ident),*) => {
        #[derive(Debug, Default, PartialEq, Eq)]
        pub struct SurfaceArena {
            $(
                $name: Arena<$tree>,
            )*
        }

        paste! {
            impl SurfaceArena {
                $(
                    pub fn [<alloc_ $name>](&mut self, $name: $tree) -> [<$tree Id>] {
                        self.$name.alloc($name)
                    }

                    pub fn [<iter_ $name>](&self) -> impl Iterator<Item = &$tree> {
                        self.$name.iter().map(|(_, tree)| tree)
                    }
                )*
            }

            $(
                impl Index<[<$tree Id>]> for SurfaceArena {
                    type Output = $tree;

                    fn index(&self, index: [<$tree Id>]) -> &$tree {
                        &self.$name[index]
                    }
                }
            )*
        }
    };
}

_surface_arena!(expr: Expr, let_name: LetName, binder: Binder, ty: Type);

// endregion

// region: SourceMap

#[derive(Debug, PartialEq, Eq, Hash)]
pub struct LetNamePtr {
    pub annotation_ptr: Option<SyntaxNodePtr>,
    pub equations_ptr: Vec<SyntaxNodePtr>,
}

/// A mapping from surface IDs to CST pointers.
#[derive(Debug, Default, PartialEq, Eq)]
pub struct SourceMap {
    pub(crate) expr_to_cst: FxHashMap<ExprId, SyntaxNodePtr>,
    pub(crate) cst_to_expr: FxHashMap<SyntaxNodePtr, ExprId>,
    pub(crate) let_name_to_cst: FxHashMap<LetNameId, LetNamePtr>,
    pub(crate) cst_to_let_name: FxHashMap<SyntaxNodePtr, LetNameId>,
    pub(crate) binder_to_cst: FxHashMap<BinderId, SyntaxNodePtr>,
    pub(crate) cst_to_binder: FxHashMap<SyntaxNodePtr, BinderId>,
    pub(crate) type_to_cst: FxHashMap<TypeId, SyntaxNodePtr>,
    pub(crate) cst_to_type: FxHashMap<SyntaxNodePtr, TypeId>,
}

// endregion
