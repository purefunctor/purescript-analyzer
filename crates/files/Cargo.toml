[package]
name = "files"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
indexmap = "2.0.0"
rustc-hash = "2.0.0"
