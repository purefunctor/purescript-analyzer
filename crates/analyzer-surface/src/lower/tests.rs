//! Snapshot tests for lowering.
//!
//! These tests ensure that CST nodes are lowered correctly as well as
//! consistently, especially for refactorings like [#9].
//!
//! [#9]: https://github.com/purefunctor/purescript-analyzer/issues/9

use std::fs;
use std::path::PathBuf;
use std::sync::Arc;

use analyzer_index::IndexStorage;
use analyzer_interner::{InternerDatabase, InternerStorage};
use analyzer_source::{SourceDatabase, SourceStorage};
use files::{ChangedFile, FileId, Files};
use itertools::Itertools;
use salsa::Durability;

use crate::tree::*;
use crate::{SurfaceDatabase, SurfaceStorage};

#[salsa::database(IndexStorage, InternerStorage, SourceStorage, SurfaceStorage)]
pub struct TestDatabase {
    storage: salsa::Storage<TestDatabase>,
}

impl salsa::Database for TestDatabase {}

impl Default for TestDatabase {
    fn default() -> TestDatabase {
        let mut db = TestDatabase { storage: Default::default() };
        db.set_interner(Default::default());
        db
    }
}

fn default_db(source: &str) -> (TestDatabase, Files, FileId) {
    let mut db = TestDatabase::default();
    let mut files = Files::default();

    files.set_file_contents("./Main.purs".into(), Some(source.into()));

    for ChangedFile { file_id, .. } in files.take_changes() {
        let contents = files.file_contents(file_id);
        db.set_file_contents(file_id, Arc::from(std::str::from_utf8(contents).unwrap()));
    }

    db.set_file_paths_with_durability(files.iter().collect(), Durability::HIGH);

    let file_id = files.file_id("./Main.purs".into()).unwrap();

    (db, files, file_id)
}

fn file_surface(source: &str) -> (Arc<Module>, Arc<SurfaceArena>) {
    let (db, _, file_id) = default_db(source);
    db.file_surface(file_id)
}

#[test_each::path(glob = "crates/analyzer-surface/src/lower/inputs/*.purs")]
fn test_file(path: PathBuf) {
    let content = fs::read_to_string(&path).unwrap();
    let (surface, arena) = file_surface(&content);

    let name = path.file_name().unwrap().to_str().unwrap();
    let expressions = arena.iter_expr().collect_vec();
    let let_names = arena.iter_let_name().collect_vec();
    let binders = arena.iter_binder().collect_vec();
    let types = arena.iter_ty().collect_vec();

    insta::assert_debug_snapshot!(name, (surface, expressions, let_names, binders, types));
}
