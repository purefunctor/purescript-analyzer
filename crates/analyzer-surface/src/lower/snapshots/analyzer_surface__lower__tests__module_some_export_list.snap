---
source: crates/analyzer-surface/src/lower/tests.rs
expression: surface.header.export_list
---
Some(
    ExportList {
        items: [
            ExportType(
                Name(
                    "Type",
                ),
                None,
            ),
            ExportType(
                Name(
                    "Data",
                ),
                Some(
                    DataAll,
                ),
            ),
            ExportType(
                Name(
                    "List",
                ),
                Some(
                    DataEnumerated(
                        [
                            Name(
                                "Cons",
                            ),
                            Name(
                                "List",
                            ),
                        ],
                    ),
                ),
            ),
            ExportValue(
                Name(
                    "value",
                ),
            ),
            ExportClass(
                Name(
                    "Class",
                ),
            ),
        ],
    },
)
